File: Pilas/pila.js
class Node{
    constructor(data){
        this.data = data;
        this.next = undefined;
    }
}

class Stack{
    constructor(){
        this.head = undefined;
    }

    isEmpty(){
        return this.head == undefined;
    }

    push(data){
        const newNode = new Node(data);
        newNode.next = this.head;
        this.head = newNode;
    }

    pop(){
        if (this.isEmpty()) return undefined;
        const poppedNode = this.head;
        this.head = this.head.next;
        poppedNode.next = undefined;
        return poppedNode;
    }

    peek(){
        if (this.isEmpty()) return undefined;
        return this.head.data;
    }

    display(){
        if (this.isEmpty()) return undefined;
        let current = this.head;
        while (current){
            console.log(current.data + '->');
            current = current.next;
        }
        return;
    }
}

const stack = new Stack();

/* Apilamos los datos dentro de la pila */
stack.push("Primer Elemento");
stack.push("Segundo Elemento");
stack.push("Tercero Elemento");
stack.push("Cuarto Elemento");

/* Mostramos el dato que esta en la cabeza */
console.log("Primer dato en salir:");
console.log(stack.peek());
console.log("-".repeat(50));

/* Mostramos todos los datos que hay dentro de la pila */
console.log("Elementos de la pila: ");
stack.display();
console.log("-".repeat(50));

/* Desapilamos dos elementos */
stack.pop();
stack.pop();

/* Mostramos la pila ahora con los elementos desapilados */
console.log("Elementos de la pila luedo de desapilarla dos veces: ");
stack.display();
console.log("-".repeat(50));

File: Pilas/pila.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
 
class Stack:
    def __init__(self):
        self.head = None
 
    def is_empty(self):
        return self.head is None
 
    def push(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node
 
    def pop(self):
        if self.is_empty():
            return None
        popped_node = self.head
        self.head = self.head.next
        popped_node.next = None
        return popped_node.data
 
    def peek(self):
        if self.is_empty():
            return None
        return self.head.data
 
    def display(self):
        current = self.head
        while current:
            print(current.data, end='->')
            current = current.next
        print()
  
stack = Stack()       
        
# Apilamos los datos dentro de la pila
stack.push("Primer Elemento")
stack.push("Segundo Elemento")
stack.push("Tercero Elemento")
stack.push("Cuarto Elemento")

# Mostramos el dato que esta en la cabeza 
print("Primer dato en salir:")
print(stack.peek())
print("-"*50)

# Mostramos todos los datos que hay dentro de la pila 
print("Elementos de la pila: ")
stack.display()
print("-"*50)

# Desapilamos dos elementos 
stack.pop()
stack.pop()

# Mostramos la pila ahora con los elementos desapilados
print("Elementos de la pila luedo de desapilarla dos veces: ")
stack.display()
print("-"*50)